src/App.tsx

```diff
@@ -1,24 +1,240 @@
+  const [error, setError] = useState<ErrorTypes | null>(null);
+  const [filter, setFilter] = useState(Filter.All);
+  const [title, setTitle] = useState('');
+  const [isAdding, setIsAdding] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [selectedTodoIds, setSelectedTodoIds] = useState<number[]>([]);
+  const [isEditing, setIsEditing] = useState(false);
+
+  const pushError = (erroType: ErrorTypes, time = 3000) => {
+    setError(erroType);
+    setTimeout(() => setError(null), time);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setAllTodos)
+      .catch(() => pushError(ErrorTypes.Load));
+  }, []);
+
+  const filterHandler = (array: Todo[], filterType: string) => {
+    switch (filterType) {
+      case Filter.Active:
+        return array.filter(item => !item.completed);
+      case Filter.Completed:
+        return array.filter(item => item.completed);
+      default:
+        return array;
+    }
+  };
+
+  const addTodoHandler = useCallback(() => {
+    if (!title.trim()) {
+      pushError(ErrorTypes.Add);
+
+      return;
+    }
+
+    setIsAdding(true);
+
+    const todo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      title,
+      userId: USER_ID,
+      completed: false,
+    });
+
+    addTodo(todo)
+      .then(result => {
+        setAllTodos(
+          prevTodos => [...prevTodos, result],
+        );
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setError(ErrorTypes.Add);
+        setTempTodo(null);
+      })
+      .finally(() => {
+        setIsAdding(false);
+        setTitle('');
+        setTempTodo(null);
+      });
+  }, [title]);
+
+  const deleteTodoHandler = useCallback((id: number) => {
+    deleteTodo(id)
+      .then(() => setAllTodos(
+        prevAllTodos => prevAllTodos.filter(prevTodo => prevTodo.id !== id),
+      ))
+      .catch(() => pushError(ErrorTypes.Delete));
+  }, []);
+
+  const activeTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Active), [allTodos],
+  );
+  const completedTodos = useMemo(
+    () => filterHandler(allTodos, Filter.Completed), [allTodos],
+  );
+  const visibleTodos = useMemo(
+    () => filterHandler(allTodos, filter), [allTodos, filter],
+  );
+
+  const deleteAllCompletedHandler = useCallback(() => {
+    completedTodos.map(
+      completedTodo => deleteTodoHandler(completedTodo.id),
+    );
+  }, [allTodos]);
+
+  const todoStatusChangeHandler = useCallback((id: number, data: boolean) => {
+    setSelectedTodoIds(prev => [...prev, id]);
+
+    completeTodo(id, data)
+      .then(() => setAllTodos(prevTodos => {
+        const allTodosCopy = [...prevTodos];
+        const changingTodo = { ...allTodosCopy.find(todo => todo.id === id) };
+
+        if (changingTodo) {
+          changingTodo.completed = data;
+        }
+
+        return allTodosCopy;
+      }))
```

Comment:
It's not working on the UI as you can notice) let me help you here
```suggestion
      .then(() => setAllTodos(prevTodos => {
        const todosCopy = JSON.parse(JSON.stringify(prevTodos));
        
        return  todosCopy.map((todo) =>  todo.id === id ? { ...todo, completed: isCompeled } : todo)
      }))
```